bitcoin/src/pow.rs

```diff
@@ -366,6 +366,56 @@ impl CompactTarget {
         Ok(Self::from_consensus(lock_time))
     }
 
+    /// Computes the [`CompactTarget`] from a difficulty adjustment.
+    ///
+    /// ref: <https://github.com/bitcoin/bitcoin/blob/master/src/pow.cpp>
+    ///
+    /// Given the previous Target, represented as a [`CompactTarget`], the difficulty is adjusted
+    /// by taking the timespan between them, and multipling the current [`CompactTarget`] by a factor
+    /// of the net timespan and expected timespan. The [`CompactTarget`] may not adjust by more than
+    /// a factor of 4, or adjust beyond the maximum threshold for the network.
+    ///
+    /// # Note
+    ///
+    /// Under the consensus rules, the different in the number of blocks between the headers does
+    /// not equate to the `difficulty_adjustment_interval` of [`Params`]. This is due to an off-by-one
+    /// error, and, the expected number of blocks in between headers is `difficulty_adjustment_interval - 1`
+    /// when calculating the difficulty adjustment.
+    ///
+    /// Take the example of the first difficulty adjustment. Block 2016 introduces a new [`CompactTarget`],
+    /// which takes the net timespan between Block 2015 and Block 0, and recomputes the difficulty.
+    ///
+    /// # Returns
+    ///
+    /// The expected [`CompactTarget`] recalculation.
+    pub fn from_next_work_required(
+        last: CompactTarget,
+        timespan: u64,
+        params: impl AsRef<Params>,
```

Comment:
What is the benefit of requiring the caller to calculate the timespan instead of taking a header like the approach in #2180 (and in Core)? Are we not pushing the off-by-one issue onto the caller who must now think about it?

The `timespan` approach definitely makes for nicer unit tests.

Reply:
Just to echo the point below, the user can still get burned with the dual header approach by just passing incorrect headers. I chewed on this for a while, and we could catch this in some instances when the `CompactTarget` in each `Header` does not match, making the return type `Result<Self, E>`, but it feels wrong to have this be fallible when it doesn't actually guarantee the headers are correct. I can definitely revert back to two headers in the signature, perhaps cleaner for integration, but the user could just pass in nonsense and get out nonsense. I went with the mantra if they read the example, the reference in cpp, and they _still_ don't understand, then they would have passed incorrect headers. 

Full disclosure I did consider the easier unit tests as well!

Reply:
Thanks man, I consider all my points fully dealt with, happy to see the next version you push.